sips/cluster_consensus.md

```diff
@@ -0,0 +1,296 @@
+	highestAttestingSlot spec.Slot
+}
+
+
+// ClusterRunner manages the duty cycle for a certain slot
+type ClusterRunner interface {
+	// Start the duty lifecycle for the given slot. Emits a message.
+    StartDuties(slot spec.Slot, attesterDuties []*types.Duty, syncCommitteeDuties []*types.Duty) error
+	// Processes cosensus message
+    ProcessConsensus(consensusMessage *qbft.Message) error
+	// Processes a post-consensus message
+	ProcessPostConsensus(msg ClusterSignaturesMessage) 
+}
+
+type ClusterRunner struct {
+	Share          *types.Share
+	QBFTController *qbft.Controller
+	BeaconNetwork  *types.BeaconNetwork
+	Signer		   *types.BeaconSigner
+}
+
+// BeaconVote is the consensus data
+type BeaconVote struct {
+    BlockRoot         phase0.Root
+    Source            phase0.Checkpoint
+    Target            phase0.Checkpoint
+}
+
+// PartialSignature holds the ValidatorIndex, Root, and Signature
+type PartialSignature struct {
+    ValidatorIndex phase0.ValidatorIndex
+    Root phase0.Root // TODO: is it really needed?
+    Signature phase0.BLSSignature
+}
+
+// ClusterSignaturesMessage holds all PartialSignatures
+type ClusterSignaturesMessage struct {
+    Signer types.OperatorID
+    Attestations []PartialSignature
+    SyncCommitteeMessages []PartialSignature
+}
+
+func ConstructAttestation(vote BeaconVote, duty AttesterDuty) Attestation {
+    bits := bitfield.New(duty.CommitteeLength)
+    bits.Set(duty.ValidatorCommitteeIndex)
+
+    return Attestation{
+        Slot: duty.Slot,
+        BlockRoot: vote.BlockRoot,
+        Source: vote.Source,
+        Target: vote.Target,
+        CommiteeIndex: duty.CommitteeIndex,
+        AggregationBits: bits,
+    }
+}
+```
+
+#### Happy Flow
+
+1. `Cluster` receives duties that match a certain slot. If the slot is higher then `highestDecidedSlot` starts consensus for the relevant roles, and initializes a `ClusterRunner` for the relevant Validators.
+2. `Cluster` receives consensus messages and hands them over `ClusterRunner` that hands them over to the `QBFTController` that has unchanged logic. The only difference is `BeaconVote` is used as the ConsensusData object.
+3.  Once `ProcessConsensus` decides, the `Cluster` will create a post consensus of PartialSignatureMessage that aggregates the beacon partial signature for all relevant validators.
+4. `Cluster` will process post-consensus partial signature messages and submit a beacon message for each validator.
+
+### Stopping Runs
+
+Previously we have letted new validator duties stop the run for the previous duty. For a cluster this is not a good idea and instead we will count on a tuned `CutOffRound` per duty to stop the instance.
+
+### Omitting Partial Signatures
+
+If in post-consensus stage for attestation duty, the duty's slot is lower than a validator's `ClusterShare's` `highestAttestingSlot`, the `ClusterRunner` will omit the post-consesnus message for this validator. 
```

Comment:
IMO this description is lacking what happens regards a quorum, and/or who created the message. if we got 2 partials and we created the third should we still omit it? if we have 2 and received a third should we omit it?

Reply:
This section only talks about omitting messages because of slots.